Learn HTML (modules 12-20) (#9576)
GoogleChrome/web.dev: src/site/content/en/learn/html/focus/index.md

authors:
  - estelleweyl
description: How to manage focus order in your HTML documents.
date: 2023-14-02
tags:
  - html
---

Interactive elements, including [form controls](/learn/html/forms/), [links](/learn/html/links/), and buttons, are by
default focusable and tabbable. Tabbable elements are part of the document's sequential focus navigation order. Other elements
are inert, meaning they are not interactive. With HTML attributes, it is possible to make interactive elements inert and to
make inert elements interactive.

{% Aside %}
For usability reasons, always ensure that the user knows which element has focus. Include CSS [`:focus`](https://developer.mozilla.org/docs/Web/CSS/:focus),
[`:focus-visible`](https://developer.mozilla.org/docs/Web/CSS/:focus-visible), and, optionally, [`:focus-within`](https://developer.mozilla.org/docs/Web/CSS/:focus-within)

> styles. This is so important: there are Learn sections devoted to focus styles in both the CSS and Accessibility learning paths.

Reviewer: ```suggestion
styles. This is so important: there are Learn sections devoted to focus styles in both the [CSS](/learn/css/focus/) and [Accessibility](/learn/accessibility/focus/) learning paths.
```

{% endAside %}

By default, the navigation focus order is the same as the visual order, which is the source code order. There are HTML
attributes that can alter this order and CSS properties that can alter the visual order of content. Changing the tabbing
order with HTML or visual rendering order with CSS can harm user experience.

Don't alter the perceived and actual tabbing order with CSS and HTML. As the following two examples demonstrate, tab orders
that differ from the visually expected order are confusing to users and bad for user experience.

In this example, the value of the [`tabindex`](https://developer.mozilla.org/docs/Web/HTML/Global_attributes/tabindex) attribute
has made the tab order chaotic:

{% Codepen {
user: 'web-dot-dev',
id: 'GRXKqGz',
tab: 'result'
} %}

In this example, CSS has created a divergence between the tabbing order and the visual order of the content:

> {% Codepen {

Reviewer: there is an issue with my original codepen. I blame the cat! Please remove the[;"]

user: 'web-dot-dev',
id: 'KKxPMBV',
tab: 'result'
} %}

The [`flex-flow: row-reverse;`](https://developer.mozilla.org/docs/Web/CSS/flex-flow) declaration has reversed the visual order.

> The CSS [order](https://developer.mozilla.org/docs/Web/CSS/order) property applied to the seventh word visually moved that
> one word. The tabbing sequence is the order of the code, which no longer matches the visual order, creating a disconnect

Reviewer: ```suggestion
In addition, the CSS [order](https://developer.mozilla.org/docs/Web/CSS/order) property was applied to the sixth word, "This", which visually moved that
one word. The tabbing sequence is the order of the code, which no longer matches the visual order, creating a disconnect
```

for keyboard users.

{% Aside %}
CSS features, including [flexbox](https://developer.mozilla.org/docs/Learn/CSS/CSS_layout/Flexbox), [grid](https://developer.mozilla.org/docs/Web/CSS/CSS_Grid_Layout),
[positioning](https://developer.mozilla.org/docs/Learn/CSS/CSS_layout/Positioning), [transforms](https://developer.mozilla.org/docs/Web/CSS/translate),
and [multi-column](https://developer.mozilla.org/docs/Learn/CSS/CSS_layout/Multiple-column_Layout), can alter the visual order of
content. Always ensure your content maintains a logical tabbing order on all viewport sizes. Test your content by tabbing
through it with a keyboard: with shift + tab to move backwards through the content. Always ensure it's apparent which element
currently has focus with CSS, and avoid accessibility problems by not reordering focusable elements with CSS.
{% endAside %}

## Making inert elements interactive

The `contentedible` and `tabindex` attributes, being global attributes, can be added to any element, making them focusable

> in the process. Focusable elements can also be focused with a mouse or pointer by having the [`autofocus`](https://developer.mozilla.org/docs/Web/HTML/Global_attributes/autofocus)

Reviewer: ```suggestion
in the process. Focusable elements can also be focused with a mouse or pointer, by having the [`autofocus`](https://developer.mozilla.org/docs/Web/HTML/Global_attributes/autofocus)
```

attribute set, or by script, such as with [`element.focus()`](https://developer.mozilla.org/docs/Web/API/HTMLElement/focus).

### The `tabindex` attribute

The global [`tabindex`](https://developer.mozilla.org/docs/Web/HTML/Global_attributes/tabindex) attribute, introduced in
[attributes](/learn/html/attributes/#tabindex), enables elements that otherwise would not be able to receive focus to get
focus, usually with the Tab key, hence the name.

The `tabindex` attribute takes as its value an integer. A negative value makes an element focusable but not tabbable. A
`tabindex` value of `0` makes the element focusable and tabbable, adding the element on which it is applied to the sequential
focus navigation order in source code order. A value of 1 or greater makes the element focusable and tabbable,
but adds it to a prioritized tabbing sequence, and, as we saw above, should be avoided.

> On this page, the share button is a custom `<share-action>` element. The `tabindex="0"` adds this not-normally focusable

Reviewer: ```suggestion
On this page, the share button, `<share-action>` is a [custom element](/learn/html/template/). The `tabindex="0"` adds this not-normally focusable
```

To re-enable a disabled element, the attribute has to be removed, generally via [`Element.removeAttribute('disabled')`](https://developer.mozilla.org/docs/Web/API/Element/removeAttribute).

The [`HTMLInputElement.disabled`](https://developer.mozilla.org/docs/Web/API/HTMLInputElement/disabled) property lets you check
if an input is disabled. As `disabled` is not a global attribute, it's not inherited from the HTMLElement, but every supporting
element interface, like [`HTMLSelectElement`](https://developer.mozilla.org/docs/Web/API/HTMLSelectElement/disabled),
[`HTMLTextareaElement`](https://developer.mozilla.org/docs/Web/API/HTMLTextAreaElement#instance_properties), has the same read-only property.

The `disabled` attribute does not apply to normally `inert` elements that are made focusable via `tabindex` or `contenteditable`.
It also doesn't apply to the `<form>` element itself. To disable these, the global `inert` attribute can be used.

### The `inert` attribute

When the `inert` global boolean attribute is added to an element, that element and all nested content become disabled—neither

> clickable nor tabbable—and removed from the accessibility tree. While inert can be applied to any element, it is generally

Reviewer: ```suggestion
clickable nor tabbable—and removed from the accessibility tree. While `inert` can be applied to any element, it is generally
```

used for sections of content, such as offscreen or otherwise hidden content.

When applying `disabled` to form controls, the browser provides default styling and can be styled using the [`:disabled`](https://developer.mozilla.org/docs/Web/CSS/:disabled)

> pseudo class. The `inert` attribute provides no visual indicators and has no matching pseudoclass (though the `[inert]` attribute selector matches).

Reviewer: ```suggestion
pseudo class. The `inert` attribute provides no visual indicators and has no matching pseudoclass (though the `[inert]` [attribute selector](/learn/css/selectors/#attribute-selector) matches).
```
